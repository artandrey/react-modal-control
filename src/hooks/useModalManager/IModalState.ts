import { ModalId, LifecycleStateLiteral } from '../../types/public-types';

export interface IModalState<P = any> {
  readonly Element: React.ComponentType<P>;
  readonly id: ModalId;
  readonly key: PropertyKey;
  readonly props: P;
  readonly lifecycleState: LifecycleStateLiteral;
}
