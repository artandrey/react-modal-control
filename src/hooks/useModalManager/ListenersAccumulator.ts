export type Listener<A, R> = (arg: A) => R;

export class ListenersAccumulator<L extends Listener<any, any>> {
  private readonly listeners: Set<L> = new Set();

  addListener(listener: L) {
    this.listeners.add(listener);
  }

  removeListener(listener: L) {
    this.listeners.delete(listener);
  }

  execute(...arg: Parameters<L>) {
    this.listeners.forEach((listener) => listener(arg));
  }
}
