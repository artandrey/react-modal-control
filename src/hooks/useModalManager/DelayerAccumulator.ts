import { generateId } from '../../helpers/generateId';
import { ListenersAccumulator } from './ListenersAccumulator';

export type DelayerListener = () => void;

export class DelayerAccumulator {
  private delayers: Set<number> = new Set();
  private listenersAccumulator = new ListenersAccumulator<DelayerListener>();
  private isActivated: boolean = false;
  private isExecuted: boolean = false;

  add() {
    const id = generateId();
    this.delayers.add(id);
    return id;
  }

  remove(delayerId: number) {
    this.delayers.delete(delayerId);
    if (this.delayers.size === 0 && this.isActivated) this.executeListeners();
  }

  addListener(onRelease: DelayerListener) {
    this.listenersAccumulator.addListener(onRelease);
  }

  removeListener(onRelease: DelayerListener) {
    this.listenersAccumulator.removeListener(onRelease);
  }

  private executeListeners() {
    if (!this.isExecuted) this.listenersAccumulator.execute();
    this.isExecuted = true;
  }

  activate() {
    this.isActivated = true;
    if (this.delayers.size === 0) this.executeListeners();
  }
}
