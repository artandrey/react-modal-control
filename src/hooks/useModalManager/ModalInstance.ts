import {
  LifecycleStateLiteral,
  LifecycleState,
  ModalId,
} from '../../types/public-types';
import { DelayerAccumulator } from './DelayerAccumulator';
import { IModalState } from './IModalState';
import { ListenersAccumulator } from './ListenersAccumulator';
import { ModalInstanceOptions } from './useModalManager';

export interface ModalInstanceLifecycleHooks {
  onCloseStart?: (instance: ModalInstance) => void;
  onClosed?: (instance: ModalInstance) => void;
  onOpen?: (instance: ModalInstance) => void;
}

export type StartCloseListener = () => void;

export class ModalInstance {
  lifecycleState: LifecycleStateLiteral = LifecycleState.PENDING;
  private readonly closeDelayers = new DelayerAccumulator();
  private readonly startCloseListenersAccumulator =
    new ListenersAccumulator<StartCloseListener>();

  constructor(
    private readonly element: React.ComponentType<any>,
    public readonly id: ModalId,
    public readonly key: PropertyKey,
    private readonly options: ModalInstanceOptions
  ) {
    this.closeDelayers.addListener(() => {
      this.endClose();
    });
  }

  public get Element() {
    return this.element;
  }

  public get props() {
    return this.options.modalOptions.props ?? {};
  }

  addCloseDelayer() {
    return this.closeDelayers.add();
  }

  removeCloseDelayer(delayerId: number) {
    this.closeDelayers.remove(delayerId);
  }

  addStartCloseListener(listener: StartCloseListener) {
    this.startCloseListenersAccumulator.addListener(listener);
  }

  removeStartCloseListener(listener: StartCloseListener) {
    this.startCloseListenersAccumulator.removeListener(listener);
  }

  startClose() {
    if (
      this.lifecycleState === LifecycleState.CLOSED ||
      this.lifecycleState !== LifecycleState.OPENED
    )
      return;
    this.lifecycleState = LifecycleState.CLOSING;
    this.executeHook('onCloseStart');
    this.startCloseListenersAccumulator.execute();
    this.closeDelayers.activate();
  }

  endClose() {
    if (this.lifecycleState === LifecycleState.CLOSED) return;
    this.lifecycleState = LifecycleState.CLOSED;
    this.executeHook('onClosed');
  }

  open() {
    if (this.lifecycleState !== LifecycleState.PENDING) return;
    this.lifecycleState = LifecycleState.OPENED;
    this.executeHook('onOpen');
  }

  private executeHook(hook: keyof ModalInstanceLifecycleHooks) {
    const callback = this.options.hooks ? this.options.hooks[hook] : null;
    if (callback) callback(this);
  }

  toState(): IModalState {
    return {
      id: this.id,
      Element: this.element,
      key: this.key,
      props: this.props,
      lifecycleState: this.lifecycleState,
    };
  }
}
