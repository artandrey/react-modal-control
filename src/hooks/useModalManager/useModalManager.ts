import { useCallback, useRef } from 'react';
import {
  AppearenceMode,
  AppearenceMods,
  LifecycleState,
  LifecycleType,
  LifecycleTypes,
  ModalId,
  ModalOpeningOptions,
  ModalWindows,
} from '../../types/public-types';
import { generateId } from '../../helpers/generateId';
import { EventStateProvider, useEventState } from '../useEventState';
import { ModalInstance, ModalInstanceLifecycleHooks } from './ModalInstance';
import { IModalState } from './IModalState';
import { getFirstMapElement } from '../../helpers/getFirstMapElement';

export interface ModalInstanceOptions {
  hooks?: ModalInstanceLifecycleHooks;
  modalOptions: ModalOpeningOptions<any>;
}

export interface LifecycleOptions {
  type: LifecycleTypes;
  appearenceMode?: AppearenceMods;
}

export interface ModalManagerOptions extends LifecycleOptions {
  modals: ModalWindows;
}

export interface ModalManager {
  open: (key: PropertyKey, options: ModalOpeningOptions<any>) => void;
  close: (id?: number) => void;
  eventProvider: EventStateProvider<IModalState[]>;
  getInstanceById(id: number): ModalInstance | undefined;
}

export function useModalManager(options: ModalManagerOptions): ModalManager {
  const { modals: modalWindows, type, appearenceMode } = options;

  const modalInstancesRef = useRef<Map<ModalId, ModalInstance>>(new Map());

  const [, setModals, eventProvider] = useEventState<IModalState[]>([]);

  const open = useCallback(
    (key: PropertyKey, options: ModalOpeningOptions<any>) => {
      const id = generateId();

      const modal = modalWindows[key];

      if (!modal)
        throw new Error(`Modal with key: ${String(key)} was not found`);

      function getFirstInQueue() {
        for (const queuedInstance of modalInstancesRef.current.values()) {
          if (queuedInstance.lifecycleState === LifecycleState.PENDING) {
            return queuedInstance;
          }
        }
      }

      const instance = new ModalInstance(modal, id, key, {
        modalOptions: options,
        hooks: {
          onClosed: (instance) => {
            setModals((modals) =>
              modals.filter((modal) => modal.id !== instance.id)
            );
            modalInstancesRef.current.delete(instance.id);

            if (
              type === LifecycleType.QUEUE &&
              appearenceMode === AppearenceMode.AFTER_CLOSE
            ) {
              getFirstInQueue()?.open();
            }
          },
          onOpen: () => {
            setModals((modals) => [...modals, instance.toState()]);
          },
          onCloseStart: (instance) => {
            setModals((modals) =>
              modals.map((modal) =>
                modal.id === instance.id ? instance.toState() : modal
              )
            );
            if (
              type === LifecycleType.QUEUE &&
              appearenceMode === AppearenceMode.DURING_CLOSE
            ) {
              getFirstInQueue()?.open();
            }
          },
        },
      });

      modalInstancesRef.current.set(instance.id, instance);

      if (
        modalInstancesRef.current.size === 1 ||
        type === LifecycleType.MULTIPLE
      ) {
        instance.open();
      }
    },
    [setModals, modalWindows, type, appearenceMode]
  );

  const close = useCallback(
    (id?: ModalId) => {
      const modals = modalInstancesRef.current;

      let modal: ModalInstance | void;

      if (id === undefined) {
        modal = getFirstMapElement(modals);
      } else {
        modal = modals.get(id);
      }

      if (!modal) return;

      modal.startClose();
    },
    [setModals]
  );

  const getInstanceById = useCallback((id: ModalId) => {
    return modalInstancesRef.current.get(id);
  }, []);

  return {
    open,
    close,
    eventProvider,
    getInstanceById,
  };
}
