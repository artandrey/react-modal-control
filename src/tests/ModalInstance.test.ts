import { vi } from 'vitest';
import { IModalState } from '../hooks/useModalManager/IModalState';
import { ModalInstance } from '../hooks/useModalManager/ModalInstance';

test('Matches state snapshot', () => {
  const Element = () => '';
  const stateSnapshot: IModalState = {
    Element: Element,
    id: 1,
    key: '',
    lifecycleState: 'PENDING',
    props: {},
  };

  const instnaceWithProps = new ModalInstance(Element, 1, '', {
    modalOptions: { props: {} },
  });

  const instanceWithoutProps = new ModalInstance(Element, 1, '', {
    modalOptions: { props: undefined },
  });

  expect(instnaceWithProps.toState()).toMatchSnapshot(stateSnapshot);
  expect(instanceWithoutProps.toState()).toMatchSnapshot(stateSnapshot);
});

function createCallbaskSpy() {
  const callbacks = {
    onOpen() {},
    onCloseStart() {},
    onClosed() {},
  };

  const onOpenSpy = vi.spyOn(callbacks, 'onOpen');
  const onCloseStartSpy = vi.spyOn(callbacks, 'onCloseStart');
  const onCloseEndSpy = vi.spyOn(callbacks, 'onClosed');

  return { callbacks, onOpenSpy, onCloseStartSpy, onCloseEndSpy };
}

describe('Testing instance hooks', () => {
  const { callbacks, onCloseEndSpy, onCloseStartSpy, onOpenSpy } =
    createCallbaskSpy();

  const instance = new ModalInstance(() => '', 1, '', {
    modalOptions: { props: {} },
    hooks: callbacks,
  });

  test('After creation any test shouldn`t be called', () => {
    expect(onOpenSpy).not.toHaveBeenCalled();
    expect(onCloseStartSpy).not.toHaveBeenCalled();
    expect(onCloseEndSpy).not.toHaveBeenCalled();
  });

  test('Open should call only open hook', () => {
    instance.open();
    instance.open();
    expect(onOpenSpy).toHaveBeenCalled();
    expect(onCloseStartSpy).not.toHaveBeenCalled();
    expect(onCloseEndSpy).not.toHaveBeenCalled();
  });

  test('Start close should call only close start hook and after close end hook', () => {
    instance.startClose();

    expect(onCloseStartSpy).toHaveBeenCalled();
    expect(onCloseEndSpy).toHaveBeenCalled();
  });

  test('After calling methods twice, hooks Open and Closed called once', () => {
    instance.open();
    instance.open();
    instance.endClose();
    instance.endClose();

    expect(onOpenSpy).toBeCalledTimes(1);
    expect(onCloseEndSpy).toBeCalledTimes(1);
  });
});
