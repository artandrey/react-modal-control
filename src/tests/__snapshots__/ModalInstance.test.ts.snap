// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Matches state snapshot 1`] = `
{
  "Element": [Function],
  "id": 1,
  "key": "",
  "lifecycleState": "PENDING",
  "props": {},
}
`;

exports[`Matches state snapshot 2`] = `
{
  "Element": [Function],
  "id": 1,
  "key": "",
  "lifecycleState": "PENDING",
  "props": {},
}
`;
